src/wasm/wasm-binary.cpp

```diff
@@ -1520,23 +1520,13 @@ void WasmBinaryWriter::writeType(Type type) {
     // internally use more refined versions of those types, but we cannot emit
     // those more refined types.
     if (!wasm->features.hasGC()) {
-      if (Type::isSubType(type, Type(HeapType::func, Nullable))) {
-        o << S32LEB(BinaryConsts::EncodedType::funcref);
-        return;
-      }
-      if (Type::isSubType(type, Type(HeapType::ext, Nullable))) {
-        o << S32LEB(BinaryConsts::EncodedType::externref);
-        return;
-      }
-      if (Type::isSubType(type, Type(HeapType::exn, Nullable))) {
-        o << S32LEB(BinaryConsts::EncodedType::exnref);
-        return;
-      }
-      if (Type::isSubType(type, Type(HeapType::string, Nullable))) {
-        o << S32LEB(BinaryConsts::EncodedType::stringref);
-        return;
+      auto ht = type.getHeapType();
+      if (ht.isBasic() && ht.getBasic(Unshared) == HeapType::string) {
+        // Do not overgeneralize stringref to anyref.
```

Comment:
Why not?

Reply:
We actually already have a couple tests for this case, and I wanted to avoid breaking them. It's admittedly a little odd (i.e. definitely wrong) that we're ok generalizing `none` to `any` even in cases where it is used for null strings, but we're not ok with generalizing stringref itself to `anyref`, but I don't think it's a big deal in practice since `stringref` effectively does not exist.

Reply:
Sounds good. Please add a comment though as the special-casing here looks surprising. lgtm with that.